### DISCUSSION TITLE
How to deal with Microsoft.Kiota.Abstractions.Serialization.Untyped* instances

### DISCUSSION THREAD

**Discussion by stefanboerner** [microsoftgraph/msgraph-sdk-dotnet#2422] [Q&A] [answered]
I'm currently struggling with reading instances of [customSecurityAttributeValue](https://learn.microsoft.com/en-us/graph/api/resources/customsecurityattributevalue?view=graph-rest-1.0) using Microsoft.Graph 5.40.0.

This property has an AdditionalData property containing a Dictionary. The Key seems to be the AttributeSet. The value is of type 
Microsoft.Kiota.Abstractions.Serialization.UntypedObject. I see all the attributes for that set in the debugger but aside from reflection I don't see a (generic) way to read all of this attributes. 
![grafik](https://github.com/microsoftgraph/msgraph-sdk-dotnet/assets/23232060/462da5fd-a903-4b65-9a66-594de881c561)

So what is the suggested generic way deserialize/read data from UntypedObject, UntypedInteger, UntypedArray etc.

**Comment by andrueastman** [comment, marked as answer]
Thanks for raising this @stefanboerner 

Ideally, the model should be defined in the metadata as type so that it shows up as a strongly typed property. However, since this is only known at runtime, you can probably parse this as below.

```cs
private static void ParseUnknownObject(UntypedNode untypedNode, string indent = "")
{
    switch (untypedNode)
    {
        case UntypedObject untypedObject:
            Console.WriteLine(indent + "Found object value: ");
            var properties = untypedObject.GetValue();
            foreach (var (name, node) in properties)
            {
                Console.WriteLine(indent + "Property Name: " + name);
                ParseUnknownObject(node, indent + "  ");
            }
            break;
        case UntypedArray untypedArray:
            Console.WriteLine(indent + "Found array value: ");
            foreach (var item in untypedArray.GetValue())
            {
                Console.WriteLine(indent + "New Item: ");
                ParseUnknownObject(item, indent + "  ");
            }
            break;
        default: //its a scalar value like a string/integer/bool
            Console.WriteLine(indent + "Found scalar value : " + KiotaJsonSerializer.SerializeAsString(untypedNode));
            break;

    }
}
```

Alternatively, if you wish to get the property as a json string representation of the object, you can also call and parse it with your favorite json parsing library. 

```cs
 var jsonString = KiotaJsonSerializer.SerializeAsString(untypedNode);
```

Lemme know if this is helpful.

**Comment by garrytrinder** [comment]
I don't have answer, but facing similar problems after upgrading from v5.43.0 to v5.48.0 (latest).

In v5.43.0, I can simply deserialize using

```chsharp
var product = JsonConvert.DeserializeObject<Product>(item.AdditionalData["fields"].ToString());
```

In v5.48.0, the above code throws an error as `item.AdditionalData["fields"].ToString()` returns an `UntypedObject` instead of a string.

So like the OP question, what is the current way to access data returned in `AdditionalData` using the latest version?

/cc @andrueastman

**Reply by stefanboerner** [reply]
I haven't tested it myself but you should be able to do it similar to the following sample:
```
Product p = null;
if(item.AdditionalData["fields"] is UntypedNode un)
{
  p = JsonConvert.DeserializeObject<Product>(KiotaJsonSerializer.SerializeAsString(un));
}
```
You need to add the [Nuget package](https://www.nuget.org/packages/Microsoft.Kiota.Abstractions) to your project.

**Comment by stefanboerner** [comment]
@andrueastman Thank you for providing the sample. I think, that access to CustomSecurityAttributes should be way easier since the definition of those attributes is pretty straightforward. 

However, I managed to read the values of the attributes to a simple data structure by looking at returned data structures and "parse" them accordingly. 

When I try to update (write) security attributes I do it vice versa that means I construct UntypedObjects/Strings/Integers etc. However, there seem to be an issue with CustomSecurityAttributes with type "Integer". Although I send an "UntypedInteger" value for a particular attribute, the following exception is thrown:

> ServiceException occured
Code: Request_BadRequest
Message: 'Integer value expected for attribute.'. ParameterName: 'Testing_SVInteger' ParameterValue: '1233'.
[ODataError]: Code: Request_BadRequest - Message: 'Integer value expected for attribute.'. ParameterName: 'Testing_SVInteger' ParameterValue: '1233'.

Screenshot of the CustomSecurityAttribute instance "retvl" used to send the patch
![grafik](https://github.com/microsoftgraph/msgraph-sdk-dotnet/assets/23232060/dcc3c14a-48e7-4f8c-8cab-8b2d5b4a3f2b)

Edit: Found the issue: i forgot to add the odata type for the integer value.
